agent/consul/fsm/commands_oss.go

```diff
@@ -428,3 +430,41 @@ func (c *FSM) applyACLPolicyDeleteOperation(buf []byte, index uint64) interface{
 
 	return c.state.ACLPolicyBatchDelete(index, req.PolicyIDs)
 }
+
+func (c *FSM) applyServiceConfigEntryOperation(buf []byte, index uint64) interface{} {
+	req := structs.ConfigEntryRequest{
+		Entry: &structs.ServiceConfigEntry{},
+	}
+	if err := structs.Decode(buf, &req); err != nil {
+		panic(fmt.Errorf("failed to decode request: %v", err))
+	}
+	return c.applyConfigEntryOperation(index, req)
+}
+
+func (c *FSM) applyProxyConfigEntryOperation(buf []byte, index uint64) interface{} {
+	req := structs.ConfigEntryRequest{
+		Entry: &structs.ProxyConfigEntry{},
+	}
+	if err := structs.Decode(buf, &req); err != nil {
+		panic(fmt.Errorf("failed to decode request: %v", err))
+	}
+	if err := c.applyConfigEntryOperation(index, req); err != nil {
+		panic(err)
+	}
+	return true
```

Comment:
Actually, having got to the bottom, I realise that this is not ideal - part of the motivation of a shared type is not having to add new Raft request types every time we make a new ConfigEntry type because each one we add is more boiler plate but also prevents downgrades so means we can't really add them except in major versions.

I guess you did this because it was hard to get the decoding types to work out right otherwise. I suggest that we collapse this back to a generic one.

I tried to suggest a work around and fell into the same rabbit hole I guess you did. In the end though there is a solution which is a bit ugly but it seems reliable enough and keeps the mess self-contained in one place: c7c2d0d9674c0541d857788fb9e44e99

The best part is that we could then just use the same for Snapshot save/restore without special cases there so there should be no where in the code except that one method that needs to switch on type for encode/decode.

What do you think?